dask_searchcv/methods.py

```diff
@@ -101,10 +102,14 @@ def _extract(self, X, y, n, is_x=True, is_train=True):
             return self.cache[n, is_x, is_train]
 
         inds = self.splits[n][0] if is_train else self.splits[n][1]
-        result = safe_indexing(X if is_x else y, inds)
-
-        if self.cache is not None:
-            self.cache[n, is_x, is_train] = result
+        post_splits = getattr(self, '_post_splits', None)
+        if post_splits:
+            result = post_splits(np.array(X)[inds])
+            self.cache[n, True, is_train] = result
```

Comment:
Does `post_split` imply that you always have a cache? If so this can be rewriting as


```python
if post_splits:
    result = ...
else:
   result = safe_indexing(...)

if self.cache is not None:
    self.cache[n, is_x, is_train] = result
```

Reply:
And what happens if a users passes `GridSearchCV(..., cache_cv=False)`?

Reply:
Good question - not sure how usage of a `post_splits` callable relates to `cache_cv=False`.  My Dataset/MLDatatset work so far considered only `cache_cv=True`.  

The idea of `post_splits` is shown more in [the CVCacheSampler](https://github.com/dask/dask-searchcv/pull/61/files#diff-1c9a50d18fb7eff16bf6a269461dcdc8R139).  It is a way of calling a `sampler` function on the argument `X` given to fit.  See the usage of `sampler` argument to `EaSearchCV` in this [elm PR 221](https://github.com/ContinuumIO/elm/blob/cv-xarray-issue-204/elm/tests/test_xarray_cross_validation.py#L152). That example uses a sampler that expects `X` to be a list of dates, and the `sampler` function determines how to build an Dataset/MLDataset for the list of dates.  In other cases, `X` could be a list of file names and the `cv` argument to `EaSearchCV` controls how to split up those file names into test / train groups.  I like the idea of a `sampler` function from cross validation / hyperparameterization of Dataset/MLDataset workflows because there is no assumption about the shapes of the DataArrays (unlike cross validation in typical `sklearn` workflows where the `cv` object is used to subset train/test row groupings of a large matrix).

TODO (for me):
 * [ ] Add a test for `cache_cv=False` with a `sampler` function - fix or document requirements there.

Reply:
More thought needs to be done on the `sampler` idea in general - currently it avoids the call to `check_consistent_length`.  The `sampler` may return `X` or an `X, y` tuple.  X may be a Dataset/MLDataset or array-like and y is currently expected to be array-like.

Reply:
I think I should just make sure I call `check_consistent_length` on samples of `X` or `(X, y`) returned by `sampler` and return `FIT_FAILURE` if there are inconsistent lengths.  I'll experiment with that today/tomorrow.